number-of-1-bits/yhkee0404.py

```diff
@@ -0,0 +1,4 @@
+class Solution:
+    @lru_cache
+    def hammingWeight(self, n: int) -> int:
+        return 1 + self.hammingWeight(n - (n & - n)) if n else 0
```

Comment:
```suggestion
    dp = [0] * (1 << 6)
    for i in range(len(dp)):
        j = i
        while j:
            dp[i] += 1
            j -= j & - j
    def hammingWeight(self, n: int) -> int:
        ans = 0
        while n:
            ans += Solution.dp[n & 0b111111]
            n >>= 6
        return ans
```

이 문제의 `Follow Up` 도 최적화를 요구하는데 댓글에서 또다른 방법을 찾아서 공유합니다: https://leetcode.com/problems/number-of-1-bits/description/comments/1882536/?parent=1568380
저는 잘 몰라서 대충 `@lru_cache`만 적어뒀었는데 더 효과적인 방법인 것 같아요.
0부터 63까지 `2^6=64` 개에 대한 정답을 전처리해 두면 그만큼 희생한 공간에 대한 Trade-Off로 시간이 6배 줄어듭니다. 2^n개에 대해서는 n배입니다.
즉 `0b000000`부터 `0b111111`까지의 정답을 상수 시간에 구할 수 있게 되므로 `n >> 1` 연산을 `logn` 번 수행하며 1의 개수를 세는 대신에 `n >> 6` 연산을 `logn / 6` 번만 수행하면서 각 Bucket의 부분 해를 더해도 정답을 알 수 있습니다.
logn <= 32 비트 수에 대한 Square Root Decomposition은 sqrt(logn) < 6 이므로 `2^6=64` 개만 전처리해도 O(logn)에서 O(sqrt(logn))으로 거의 제곱만큼 시간을 줄이는 효과가 생깁니다.
나아가 저장 공간을 2배씩 더 소모하는 만큼 연산 시간의 분모에는 1씩 더하는 효과가 있네요.
O(hammingWeight(n))은 최대 32번 수행할 수도 있는 반면에 O(sqrt(logn))만 해도 최대 6번밖에 수행하지 않으므로 최악의 시간 복잡도에서 더 유리한 Bucket 알고리즘이 되네요.